Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm

```diff
@@ -373,89 +373,107 @@ static void bufferWasConsumedCallback(CMNotificationCenterRef, const void* liste
+                    trackDidChangeSelected(trackRef, selected);
+                };
+                if (initialized)
+                    videoTrackSelectedChanged();
+                else
+                    m_pendingTrackChangeTasks.append(WTFMove(videoTrackSelectedChanged));
+            });
+            continue;
+        }
+        if (auto audioTrack = dynamicDowncast<AudioTrackPrivate>(track)) {
+            audioTrack->setEnabledChangedCallback([weakThis = WeakPtr { *this }, this, initialized] (AudioTrackPrivate& track, bool enabled) {
+                if (!weakThis)
+                    return;
 
-    if (auto player = this->player(); player && player->shouldCheckHardwareSupport()) {
-        for (auto& info : segment.videoTracks) {
-            auto codec = FourCC::fromString(info.description->codec());
-            if (!codec)
-                continue;
-            if (!codecsMeetHardwareDecodeRequirements({{ *codec }}, player->mediaContentTypesRequiringHardwareSupport())) {
-                m_parsingSucceeded = false;
-                return;
-            }
+                auto audioTrackEnabledChanged = [weakThis, this, trackRef = Ref { track }, enabled] {
+                    if (!weakThis)
+                        return;
+                    trackDidChangeEnabled(trackRef, enabled);
+                };
+                if (initialized)
+                    audioTrackEnabledChanged();
+                else
+                    m_pendingTrackChangeTasks.append(WTFMove(audioTrackEnabledChanged));
+            });
         }
     }
+}
 
-    m_protectedTrackInitDataMap.swap(m_pendingProtectedTrackInitDataMap);
-    m_pendingProtectedTrackInitDataMap.clear();
+void SourceBufferPrivateAVFObjC::didParseInitializationData(InitializationSegment&& segment)
+{
+    ALWAYS_LOG(LOGIDENTIFIER);
 
-    for (auto videoTrackInfo : segment.videoTracks) {
-        videoTrackInfo.track->setSelectedChangedCallback([weakThis = WeakPtr { *this }, this] (VideoTrackPrivate& track, bool selected) {
-            if (!weakThis)
-                return;
+    Vector<RefPtr<TrackPrivateBase>> tracks;
+    for (auto& videoTrackInfo : segment.videoTracks)
+        tracks.append(videoTrackInfo.track);
 
-            auto videoTrackSelectedChanged = [weakThis, this, trackRef = Ref { track }, selected] {
-                if (!weakThis)
-                    return;
-                trackDidChangeSelected(trackRef, selected);
-            };
-
-            if (!processingInitializationSegment()) {
-                videoTrackSelectedChanged();
-                return;
-            }
+    for (auto& audioTrackInfo : segment.audioTracks)
+        tracks.append(audioTrackInfo.track);
 
-            m_pendingTrackChangeTasks.append(WTFMove(videoTrackSelectedChanged));
-        });
+    auto initCheckTask = [weakThis = WeakPtr { *this }, this, tracks, abortCalled = m_abortCalled] (InitializationSegment& segment) {
+        if (!weakThis)
```

Comment:
This lambda is doing quite some stuff, I wonder whether we should not do:
```
RefPtr protectedThis = weakThis.get();
if (!protectedThis)
    return;
```

Applicable elsewhere

Reply:
at this stage, it's all in the main thread. It can't change after it's been tested.

Reply:
'this' can be destroyed synchronously in lambdas. Hence the safer protectedThis pattern here